File: Ahorn/entities/customFeather.jl
module FrostHelperCustomFeatherModule

using ..Ahorn, Maple

@mapdef Entity "FrostHelper/CustomFeather" FrostHelperCustomFeatherPlacement(x::Integer, y::Integer, shielded::Bool=false, singleUse::Bool=false, respawnTime::Number=3.0, flyColor::String="ffd65c", spriteColor="White", flyTime::Number=2.0, maxSpeed::Number=190.0, lowSpeed::Number=140.0, neutralSpeed::Number=91.0, spritePath::String="objects/flyFeather/")

const colors = sort(collect(keys(Ahorn.XNAColors.colors)))

const placements = Ahorn.PlacementDict(
	"Custom Feather (Frost Helper)" => Ahorn.EntityPlacement(
		FrostHelperCustomFeatherPlacement
	)
)

Ahorn.editingOptions(entity::FrostHelperCustomFeatherPlacement) = Dict{String, Any}(
    "flyColor" => colors,
	"spriteColor" => colors
)

sprite = "objects/flyFeather/idle00.png"

function Ahorn.selection(entity::FrostHelperCustomFeatherPlacement)
    x, y = Ahorn.position(entity)
    return Ahorn.getSpriteRectangle(sprite, x, y)
end

Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::FrostHelperCustomFeatherPlacement, room::Maple.Room) = Ahorn.drawSprite(ctx, sprite, 0, 0)

end

File: Loenn/entities/customFeather.lua
---@module 'jautils'
local jautils = require("mods").requireFromPlugin("libraries.jautils")
local utils = require("utils")
local fallback = "objects/flyFeather/idle00"

---@type EntityHandler<UnknownEntity>
local feather = {}

feather.name = "FrostHelper/CustomFeather"
feather.depth = 0

local aimInvertions = {
    "None",
    "InvertX",
    "InvertY",
    "InvertBoth",
}

jautils.createPlacementsPreserveOrder(feather, "normal", {
    { "flyColor", "ffd65c", "color" },
    { "spriteColor", "ffffff", "color" },
    { "flyTime", 2.0 },
    { "respawnTime", 3.0 },
    { "maxSpeed", 190.0 },
    { "lowSpeed", 140.0 },
    { "neutralSpeed", 91.0 },
    { "spritePath", "objects/flyFeather/" },
    { "hitbox", "R,20,20,-10,-10", "FrostHelper.collider" },
    { "invertAim", "None", aimInvertions },
    { "version", 1, jautils.fields.integer{}, nil, { hidden = true } },
    { "shielded", false },
    { "singleUse", false },
    { "refillStamina", true },
    { "refillDashes", true },
})

function feather.sprite(room, entity)
    return jautils.union(
        jautils.getCustomSprite(entity, "spritePath", "idle00", fallback, "spriteColor"),
        entity.shielded and jautils.getCircleSprite(entity.x, entity.y, 12) or nil
    )
end

function feather.selection(room, entity)
    return utils.rectangle(entity.x - 12, entity.y - 12, 24, 24)
end

return feather